lib/src/common/search_service.dart

```diff
@@ -47,7 +47,100 @@ class SearchService extends Service {
     return controller.stream;
   }
 
-  // TODO: Implement code: https://developer.github.com/v3/search/#search-code
+  /// Search through code for a given [query]
+  /// You can include any github qualifiers in the query directly
+  /// or you can set some of the optional params to set the qualifiers
+  /// For example, these do the same thing:
+  /// code('awesome language:dart') and
+  /// code('awesome', language: 'dart')
+  ///
+  /// https://developer.github.com/v3/search/#search-code
+  Future<CodeSearchResults> code(String query,
+      {String language,
+      String filename,
+      String extension,
+      String user,
+      String org,
+      String repo,
+      String fork,
+      String path,
+      String size,
+      bool inFile: true,
+      bool inPath: false,
+      int pages: 2,
+      int perPage: 30}) async {
+    // Add qualifiers to the query
+    // Known Issue: If a query already has a qualifier and the same
+    // qualifier parameter is passed in, it will be duplicated.
+    // Example: code('example repo:ex', repo: 'ex') will result in
+    // a query of "example repo:ex repo:ex"
+    query += _searchQualifier('language', language);
+    query += _searchQualifier('filename', filename);
+    query += _searchQualifier('extension', extension);
+    query += _searchQualifier('user', user);
+    query += _searchQualifier('org', org);
+    query += _searchQualifier('repo', repo);
+    query += _searchQualifier('fork', fork);
+    query += _searchQualifier('path', path);
+    query += _searchQualifier('size', size);
+
+    // build up the in: qualifier based on the 2 booleans
+    String _in = '';
+    if (inFile) {
+      _in = 'file';
+    }
+    if (inPath) {
+      if (_in.isEmpty) {
+        _in = 'path';
+      } else {
+        _in = 'file,path';
+      }
+    }
+    if (_in.isNotEmpty) {
+      query += ' in:$_in';
+    }
+    var params = <String, dynamic>{};
+    params['q'] = query;
+    params['per_page'] = perPage?.toString();
+
+    var results = <CodeSearchResults>[];
+    Completer<CodeSearchResults> c = new Completer<CodeSearchResults>();
+
+    Stream<http.Response> responseStream = new PaginationHelper(_github)
+        .fetchStreamed("GET", "/search/code", params: params, pages: pages);
+
+    responseStream.handleError((err) {
+      if (err != null && err.toString().contains('rate limit exceeded')) {
+        throw new RateLimitHit(_github);
+      }
+    });
+
+    responseStream.listen((response) {
+      var input = json.decode(response.body);
+      if (input['items'] == null) {
+        return;
+      }
+      results.add(CodeSearchResults.fromJson(input));
+    }).onDone(() {
+      // Once we're done, combine all of the items from each page response
+      // into the first CodeSearchResults and complete with the combined results
+      if (results.length > 1) {
+        for (int i = 1; i < results.length; i++) {
+          results[0].items.addAll(results[i].items);
```

Comment:
I don't know much about the context, but this doesn't look right..?
You have a list of CodeSearchResults and you always add them to the same index. Does that mean we could return a single CodeSearchResult?

Reply:
Oh, you know what .. your question made me think more about this. Having it iterate and accumulate the responses internally means it will eagerly fetch all the pages. To be more consistent with other parts of the API, I think the code search should return `Stream<CodeSearchResults>` instead and leave it up to the consumer to pull as many responses from the stream as they want.

Reply:
That's what I was thinking about. I was gonna write an example of how I'd see this done, but yeah.